File: nodejs-compressor/src/parser/nodes/ReferenceNode.js
/**
 * @fileoverview Reference AST Node for ASON 2.0
 *
 * Represents references to defined values ($var) or objects (&obj).
 *
 * @module ReferenceNode
 * @license MIT
 * @version 2.0.0
 */

import { ASTNode } from './ASTNode.js';

/**
 * AST node representing a reference.
 *
 * References point to values defined in the $def: section:
 * - $var: named variable reference
 * - &obj: object alias reference
 * - #N: numeric reference (legacy, deprecated)
 *
 * @class ReferenceNode
 * @extends ASTNode
 */
export class ReferenceNode extends ASTNode {
  /**
   * Creates a reference node.
   *
   * @constructor
   * @param {string} name - Reference name (with prefix: $var, &obj, #0)
   * @param {'var'|'object'|'numeric'} refType - Type of reference
   * @param {Object} [metadata={}] - Optional metadata
   *
   * @example
   * const ref = new ReferenceNode('$email', 'var');
   */
  constructor(name, refType, metadata = {}) {
    super('Reference', metadata);

    /** @type {string} Full reference name (with prefix) */
    this.name = name;

    /** @type {'var'|'object'|'numeric'} Reference type */
    this.refType = refType;

    /** @type {ASTNode|null} Resolved value (set during resolution) */
    this.resolved = null;

    /** @type {boolean} Whether reference has been resolved */
    this.isResolved = false;
  }

  /**
   * Gets the reference name without prefix.
   *
   * @returns {string} Name without prefix
   *
   * @example
   * new ReferenceNode('$email', 'var').getBaseName() // 'email'
   * new ReferenceNode('&obj0', 'object').getBaseName() // 'obj0'
   */
  getBaseName() {
    return this.name.substring(1); // Remove first character (prefix)
  }

  /**
   * Gets the reference prefix.
   *
   * @returns {string} Prefix character ($, &, or #)
   */
  getPrefix() {
    return this.name.charAt(0);
  }

  /**
   * Resolves this reference to a value.
   *
   * @param {ASTNode} value - Resolved value
   */
  resolve(value) {
    this.resolved = value;
    this.isResolved = true;
  }

  /**
   * Checks if this is a variable reference ($var).
   *
   * @returns {boolean} True if variable reference
   */
  isVariableRef() {
    return this.refType === 'var';
  }

  /**
   * Checks if this is an object reference (&obj).
   *
   * @returns {boolean} True if object reference
   */
  isObjectRef() {
    return this.refType === 'object';
  }

  /**
   * Checks if this is a numeric reference (#N).
   *
   * @returns {boolean} True if numeric reference
   */
  isNumericRef() {
    return this.refType === 'numeric';
  }

  accept(visitor) {
    return visitor.visitReference?.(this) ?? this.toValue();
  }

  /**
   * Converts reference to its resolved value.
   *
   * @throws {Error} If reference is not resolved
   * @returns {*} Resolved value
   */
  toValue() {
    if (!this.isResolved) {
      throw new Error(`Unresolved reference: ${this.name}`);
    }
    return this.resolved.toValue();
  }

  toString(indent = 0) {
    const spaces = '  '.repeat(indent);
    const status = this.isResolved ? ' (resolved)' : ' (unresolved)';
    return `${spaces}Reference(${this.name})${status}`;
  }

  toJSON() {
    return {
      ...super.toJSON(),
      name: this.name,
      refType: this.refType,
      isResolved: this.isResolved,
      resolved: this.isResolved ? this.resolved.toJSON() : null
    };
  }
}

/**
 * AST node representing a definition ($def: section).
 *
 * Contains all reference definitions for the document.
 *
 * @class DefinitionNode
 * @extends ASTNode
 */
export class DefinitionNode extends ASTNode {
  /**
   * Creates a definition node.
   *
   * @constructor
   * @param {Object} [metadata={}] - Optional metadata
   */
  constructor(metadata = {}) {
    super('Definition', metadata);

    /** @type {Map<string, ASTNode>} Variable definitions ($var) */
    this.variables = new Map();

    /** @type {Map<string, ASTNode>} Object definitions (&obj) */
    this.objects = new Map();

    /** @type {Map<string, ASTNode>} Numeric definitions (#N) - legacy */
    this.numeric = new Map();
  }

  /**
   * Adds a variable definition.
   *
   * @param {string} name - Variable name (with $ prefix)
   * @param {ASTNode} value - Variable value
   */
  defineVariable(name, value) {
    this.variables.set(name, value);
  }

  /**
   * Adds an object definition.
   *
   * @param {string} name - Object name (with & prefix)
   * @param {ASTNode} value - Object value
   */
  defineObject(name, value) {
    this.objects.set(name, value);
  }

  /**
   * Adds a numeric definition.
   *
   * @param {string} name - Numeric name (with # prefix)
   * @param {ASTNode} value - Value
   */
  defineNumeric(name, value) {
    this.numeric.set(name, value);
  }

  /**
   * Looks up a reference by name.
   *
   * @param {string} name - Reference name (with prefix)
   * @returns {ASTNode|undefined} Defined value or undefined
   */
  lookup(name) {
    const prefix = name.charAt(0);

    if (prefix === '$') {
      return this.variables.get(name);
    } else if (prefix === '&') {
      return this.objects.get(name);
    } else if (prefix === '#') {
      return this.numeric.get(name);
    }

    return undefined;
  }

  /**
   * Gets all definitions.
   *
   * @returns {Map<string, ASTNode>} All definitions
   */
  getAllDefinitions() {
    const all = new Map();
    for (const [k, v] of this.variables) all.set(k, v);
    for (const [k, v] of this.objects) all.set(k, v);
    for (const [k, v] of this.numeric) all.set(k, v);
    return all;
  }

  accept(visitor) {
    return visitor.visitDefinition?.(this) ?? this.toValue();
  }

  toValue() {
    // Definitions don't have a direct value representation
    return null;
  }

  toString(indent = 0) {
    const spaces = '  '.repeat(indent);
    let str = `${spaces}Definition {\n`;

    if (this.variables.size > 0) {
      str += `${spaces}  Variables:\n`;
      for (const [name, value] of this.variables) {
        str += `${spaces}    ${name}: ${value.toString(indent + 2).trim()}\n`;
      }
    }

    if (this.objects.size > 0) {
      str += `${spaces}  Objects:\n`;
      for (const [name, value] of this.objects) {
        str += `${spaces}    ${name}: ${value.toString(indent + 2).trim()}\n`;
      }
    }

    str += `${spaces}}`;
    return str;
  }

  toJSON() {
    const vars = {};
    const objs = {};

    for (const [k, v] of this.variables) vars[k] = v.toJSON();
    for (const [k, v] of this.objects) objs[k] = v.toJSON();

    return {
      ...super.toJSON(),
      variables: vars,
      objects: objs
    };
  }
}

File: python-compressor/src/ason/parser/reference_node.py
"""
Reference AST Node for ASON 2.0
"""

from typing import Any, Dict, Optional
from .ast_nodes import ASTNode


class ReferenceNode(ASTNode):
    """AST node representing a reference ($var, &obj, #N)."""
    
    def __init__(self, name: str, ref_type: str, metadata: Optional[Dict[str, Any]] = None):
        super().__init__('Reference', metadata)
        self.name = name
        self.ref_type = ref_type
        self.resolved: Optional[ASTNode] = None
        self.is_resolved = False
    
    def get_base_name(self) -> str:
        """Get reference name without prefix."""
        return self.name[1:]  # Remove first character (prefix)
    
    def get_prefix(self) -> str:
        """Get the reference prefix."""
        return self.name[0]
    
    def resolve(self, value: ASTNode) -> None:
        """Resolve this reference to a value."""
        self.resolved = value
        self.is_resolved = True
    
    def is_variable_ref(self) -> bool:
        return self.ref_type == 'var'
    
    def is_object_ref(self) -> bool:
        return self.ref_type == 'object'
    
    def is_numeric_ref(self) -> bool:
        return self.ref_type == 'numeric'

    def to_value(self) -> Any:
        if not self.is_resolved:
            raise ValueError(f"Unresolved reference: {self.name}")
        return self.resolved.to_value()
    
    def to_string(self, indent: int = 0) -> str:
        spaces = '  ' * indent
        status = ' (resolved)' if self.is_resolved else ' (unresolved)'
        return f"{spaces}Reference({self.name}){status}"


class DefinitionNode(ASTNode):
    """AST node representing definitions ($def: section)."""

    def __init__(self, metadata: Optional[Dict[str, Any]] = None):
        super().__init__('Definition', metadata)
        self.variables: Dict[str, ASTNode] = {}
        self.objects: Dict[str, ASTNode] = {}
        self.numeric: Dict[str, ASTNode] = {}
    
    def define_variable(self, name: str, value: ASTNode) -> None:
        self.variables[name] = value

    def define_object(self, name: str, value: ASTNode) -> None:
        self.objects[name] = value

    def define_numeric(self, name: str, value: ASTNode) -> None:
        self.numeric[name] = value
    
    def lookup(self, name: str) -> Optional[ASTNode]:
        prefix = name[0] if name else ''
        
        if prefix == '$':
            return self.variables.get(name)
        elif prefix == '&':
            return self.objects.get(name)
        elif prefix == '#':
            return self.numeric.get(name)
        
        return None

    def to_value(self) -> None:
        return None
    
    def to_string(self, indent: int = 0) -> str:
        spaces = '  ' * indent
        result = f"{spaces}Definition {{\n"
        
        if self.variables:
            result += f"{spaces}  Variables:\n"
            for name, value in self.variables.items():
                result += f"{spaces}    {name}: {value.to_string(indent + 2).strip()}\n"
        
        result += f"{spaces}}}"
        return result
